### DISCUSSION TITLE
Publishing Provider packages to Dockerhub (for testing or local development)

### DISCUSSION THREAD

**Discussion by turkenh** [crossplane/crossplane#3817] [Knowledge base]
```bash
DOCKERHUB_ORG=<your-dockerhub-org>

export REGISTRY_ORGS=""
export XPKG_REG_ORGS_NO_PROMOTE=""
export XPKG_REG_ORGS="index.docker.io/$DOCKERHUB_ORG"

make build.all publish BRANCH_NAME=main
```

**Comment by turkenf** [comment]
### Publishing Provider packages to Dockerhub for Official Family Providers
- go to the directory containing the changes you want to push(for "provider-aws, provider-azure, provider-gcp")
- make sure you don't have any uncommitted changes
- run `export XPKG_REG_ORGS="index.docker.io/<your-dockerhub-org>"`

*For instance for provider-family-azure you need `alertsmanagement`, `analysisservices`, `apimanagement` packages:*

- run `make build.all publish BRANCH_NAME=main SUBPACKAGES="alertsmanagement analysisservices apimanagement"`
- go to https://hub.docker.com/ and login
![Screenshot 2023-07-12 at 17 30 37](https://github.com/crossplane/crossplane/assets/103541666/636478c2-4d84-4e0b-9618-2d67bea5fb83)
- and you can install the related image to your cluster
![Screenshot 2023-07-12 at 17 32 56](https://github.com/crossplane/crossplane/assets/103541666/0e01179c-5d76-494a-8cab-9aa9e23d435f)

**Comment by sttts** [comment]
Why is that not part of the README.md of providers?

**Reply by turkenh** [reply]
Because we have a lot of providers 😅 
But yes, this could definitely be added.

**Comment by sttts** [comment]
The `REGISTRY_ORGS="" in the first comment breaks building of family providers and stops pushing.

**Reply by turkenh** [reply]
The first one does not work with family, the second one does.

**Comment by ctenberge-oneid** [comment]
I am following as described above but with provider-upjet-aws
I have ran `make clean` then `make build.all publish BRANCH_NAME=main SUBPACKAGES="efs"`
but seeing this error 
```
09:47:17 [ OK ] docker build build-c7a357b6/provider-aws-arm64
All YAML files processed and saved to: /Users/calebtenberge/workspace/provider-upjet-aws/.work/xpkg-cleaned-examples
09:47:20 [ .. ] Batch processing smaller provider packages for: config efs
Publishing of smaller provider package has failed for service "config": failed to build smaller provider package: config: failed to get config file from image: Error response from daemon: NotFound: snapshot sha256:42252570830adccc0d70e30f5080c5ba03fa85ae979261acf2dc2b03bd26f494 does not exist: not found

Publishing of smaller provider package has failed for service "efs": failed to build smaller provider package: efs: failed to get config file from image: Error response from daemon: NotFound: snapshot sha256:42252570830adccc0d70e30f5080c5ba03fa85ae979261acf2dc2b03bd26f494 does not exist: not found

up: error: xpkg.batchCmd.Run(): processing of at least one smaller provider has failed: 
           failed to process smaller provider package for "efs": failed to build smaller provider package: efs: failed to get config file from image: Error response from daemon: NotFound: snapshot sha256:42252570830adccc0d70e30f5080c5ba03fa85ae979261acf2dc2b03bd26f494 does not exist: not found: 
           failed to process smaller provider package for "config": failed to build smaller provider package: config: failed to get config file from image: Error response from daemon: NotFound: snapshot sha256:42252570830adccc0d70e30f5080c5ba03fa85ae979261acf2dc2b03bd26f494 does not exist: not found
09:47:21 [FAIL]
make[2]: *** [batch-process] Error 1
make[1]: *** [img.release.publish.xpkg.upbound.io/upbound.provider-aws] Error 2
make: *** [publish] Error 2
```
